File: mobile/ios/App/App/MediaButtonsPlugin.swift
import Foundation
import AVFoundation
import Capacitor
import MediaPlayer
import UIKit

@objc(MediaButtonsPlugin)
public class MediaButtonsPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "MediaButtonsPlugin"
    public let jsName = "MediaButtons"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "startListening", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopListening", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "dispatchButton", returnType: CAPPluginReturnPromise)
    ]

    private let buttonTrackNext = "media-track-next"
    private let buttonTrackPrevious = "media-track-previous"
    private var isListening = false
    private var audioEngine: AVAudioEngine?
    private var silentPlayer: AVAudioPlayerNode?
    private var nextTrackTarget: Any?
    private var previousTrackTarget: Any?

    @objc func startListening(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            do {
                try self.configureRemoteCommandsIfNeeded()
                call.resolve()
            } catch {
                call.reject("Failed to start media button listening: \(error.localizedDescription)")
            }
        }
    }

    @objc func stopListening(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            self.stopRemoteCommands()
            call.resolve()
        }
    }

    /**
     * Dispatches a media button event to JavaScript listeners.
     */
    @objc func dispatchButton(_ call: CAPPluginCall) {
        guard let buttonId = call.getString("buttonId"), !buttonId.isEmpty else {
            call.reject("buttonId is required")
            return
        }

        guard isValidButtonId(buttonId) else {
            call.reject("Invalid buttonId: \(buttonId)")
            return
        }

        emitButton(buttonId)

        call.resolve()
    }

    private func configureRemoteCommandsIfNeeded() throws {
        if isListening {
            return
        }

        let session = AVAudioSession.sharedInstance()
        try session.setCategory(.playback, mode: .default, options: [.mixWithOthers])
        try session.setActive(true)

        try startSilentPlaybackIfNeeded()
        UIApplication.shared.beginReceivingRemoteControlEvents()

        MPNowPlayingInfoCenter.default().nowPlayingInfo = [
            MPMediaItemPropertyTitle: "Padel Buddy",
            MPMediaItemPropertyArtist: "Match Remote",
            MPMediaItemPropertyAlbumTitle: "Score Control",
            MPNowPlayingInfoPropertyPlaybackRate: 1.0
        ]

        let commandCenter = MPRemoteCommandCenter.shared()
        commandCenter.nextTrackCommand.isEnabled = true
        commandCenter.previousTrackCommand.isEnabled = true

        nextTrackTarget = commandCenter.nextTrackCommand.addTarget { [weak self] _ in
            self?.emitButton(self?.buttonTrackNext ?? "")
            return .success
        }

        previousTrackTarget = commandCenter.previousTrackCommand.addTarget { [weak self] _ in
            self?.emitButton(self?.buttonTrackPrevious ?? "")
            return .success
        }

        isListening = true
    }

    private func stopRemoteCommands() {
        if !isListening {
            return
        }

        let commandCenter = MPRemoteCommandCenter.shared()

        if let nextTrackTarget {
            commandCenter.nextTrackCommand.removeTarget(nextTrackTarget)
        }

        if let previousTrackTarget {
            commandCenter.previousTrackCommand.removeTarget(previousTrackTarget)
        }

        nextTrackTarget = nil
        previousTrackTarget = nil
        commandCenter.nextTrackCommand.isEnabled = false
        commandCenter.previousTrackCommand.isEnabled = false

        silentPlayer?.stop()
        audioEngine?.stop()
        silentPlayer = nil
        audioEngine = nil

        MPNowPlayingInfoCenter.default().nowPlayingInfo = nil
        UIApplication.shared.endReceivingRemoteControlEvents()
        try? AVAudioSession.sharedInstance().setActive(false, options: [.notifyOthersOnDeactivation])

        isListening = false
    }

    private func startSilentPlaybackIfNeeded() throws {
        if audioEngine != nil {
            return
        }

        let engine = AVAudioEngine()
        let player = AVAudioPlayerNode()
        let sampleRate = 44_100.0
        let frameCount = AVAudioFrameCount(sampleRate / 4)

        guard let format = AVAudioFormat(standardFormatWithSampleRate: sampleRate, channels: 1),
              let buffer = AVAudioPCMBuffer(pcmFormat: format, frameCapacity: frameCount),
              let channelData = buffer.floatChannelData else {
            throw NSError(domain: "MediaButtons", code: -1, userInfo: [
                NSLocalizedDescriptionKey: "Unable to create silent audio buffer"
            ])
        }

        buffer.frameLength = frameCount

        for frameIndex in 0 ..< Int(frameCount) {
            channelData[0][frameIndex] = 0
        }

        engine.attach(player)
        engine.connect(player, to: engine.mainMixerNode, format: format)
        engine.mainMixerNode.outputVolume = 0
        engine.prepare()
        try engine.start()

        player.scheduleBuffer(buffer, at: nil, options: [.loops], completionHandler: nil)
        player.play()

        audioEngine = engine
        silentPlayer = player
    }

    private func emitButton(_ buttonId: String) {
        guard !buttonId.isEmpty else {
            return
        }

        let eventData: [String: Any] = ["buttonId": buttonId]
        notifyListeners("mediaButton", data: eventData)
    }

    private func isValidButtonId(_ buttonId: String) -> Bool {
        return buttonId == buttonTrackNext
            || buttonId == buttonTrackPrevious
    }
}

File: mobile/android/app/src/main/java/com/padelbuddy/web/MediaButtonsPlugin.java
package com.padelbuddy.web;

import android.view.KeyEvent;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "MediaButtons")
public class MediaButtonsPlugin extends Plugin {
    private static volatile MediaButtonsPlugin activeInstance;

    /**
     * Supported button IDs that can be dispatched from native.
     */
    private static final String BUTTON_TRACK_NEXT = "media-track-next";
    private static final String BUTTON_TRACK_PREVIOUS = "media-track-previous";

    @Override
    public void load() {
        super.load();
        activeInstance = this;
    }

    @Override
    protected void handleOnDestroy() {
        if (activeInstance == this) {
            activeInstance = null;
        }

        super.handleOnDestroy();
    }

    @PluginMethod
    public void startListening(PluginCall call) {
        call.resolve();
    }

    @PluginMethod
    public void stopListening(PluginCall call) {
        call.resolve();
    }

    @PluginMethod
    public void dispatchButton(PluginCall call) {
        String buttonId = call.getString("buttonId");

        if (buttonId == null || buttonId.isEmpty()) {
            call.reject("buttonId is required");
            return;
        }

        // Validate button ID
        if (!isValidButtonId(buttonId)) {
            call.reject("Invalid buttonId: " + buttonId);
            return;
        }

        emitButton(buttonId);

        call.resolve();
    }

    public static boolean handleHardwareKey(int keyCode) {
        MediaButtonsPlugin plugin = activeInstance;

        if (plugin == null) {
            return false;
        }

        String buttonId = plugin.buttonIdFromKeyCode(keyCode);

        if (buttonId == null) {
            return false;
        }

        plugin.emitButton(buttonId);
        return true;
    }

    private void emitButton(String buttonId) {
        JSObject eventData = new JSObject();
        eventData.put("buttonId", buttonId);
        notifyListeners("mediaButton", eventData);
    }

    private String buttonIdFromKeyCode(int keyCode) {
        switch (keyCode) {
            case KeyEvent.KEYCODE_MEDIA_NEXT:
                return BUTTON_TRACK_NEXT;
            case KeyEvent.KEYCODE_MEDIA_PREVIOUS:
                return BUTTON_TRACK_PREVIOUS;
            default:
                return null;
        }
    }

    private boolean isValidButtonId(String buttonId) {
        return BUTTON_TRACK_NEXT.equals(buttonId)
            || BUTTON_TRACK_PREVIOUS.equals(buttonId);
    }
}
